ESQL: Infer not null for aggregated fields

costin opened the pull request:
Since the majority of aggregations discard null values, filter them to
 begin with.
 | STATS x = min(a), y = sum(b) BY c
 becomes
 | WHERE a IS NOT NULL OR b IS NOT NULL
 | STATS x = min(a), y = sum(b) BY c

Fix #102787

costin committed:
ESQL: Infer not null for aggregated fields

Since the majority of aggregations discard null values, filter them to
 begin with.
 | STATS x = min(a), y = sum(b) BY c
 becomes
 | WHERE a IS NOT NULL OR b IS NOT NULL
 | STATS x = min(a), y = sum(b) BY c

Fix #102787

costin committed:
Update docs/changelog/103030.yaml

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

costin committed:
More updates and discovering the filter is not applicable...

costin committed:
Restrict rule to work only on non-grouped aggregations

Currently two CSV tests are failing since the optimization kicks in but
 the propagation of the LocalRelation expect to be in the middle of a
 pipeline breaker and thus creates an intermediate output which ends up
 failing

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -1137,6 +1144,53 @@ private LogicalPlan rule(Eval eval) {
+     * to begin with.
+     * STATS x = min(a), y = sum(b)
+     * becomes
+     * | WHERE a IS NOT NULL OR b IS NOT NULL
```
If we aren't pushing down the not-null constraints we shouldn't make the EVAL bit. Or we can make it, but then it should get folded into the agg. The agg ignores the null all on it's own.

costin replied:
Moved the filter to the local optimizer so it gets added only when there's a local agg - otherwise it gets skipped.
This shows up in the tests where if an agg is on the coordinator (due to a limit for example), the rule doesn't kick in.

costin committed:
wip

costin committed:
Merge branch 'main' into fix/102787

costin committed:
ESQL: Improve pushdown of certain filters

The check of what filters can be pushed down was too strict leading to
 some queries to be evaluated outside the storage. This PR addresses the
 issue in particular is not null and is null filters.

Fix #103536

costin committed:
ESQL: Improve pushdown of certain filters

The check of what filters can be pushed down was too strict leading to
 some queries to be evaluated outside the storage. This PR addresses the
 issue in particular is not null and is null filters.

Fix #103536

costin committed:
Update docs/changelog/103538.yaml

costin committed:
Update tests and incorporate #103536

Merge branch 'fix/103536' into fix/102787

costin committed:
Move not null filter to the local optimizer as it adds value only when
 it can be pushed down

costin commented:
@elasticsearchmachine update branch

costin commented:
@elasticsearchmachine merge upstream

costin committed:
Merge branch 'main' into fix/102787

costin committed:
Update test

costin commented:
@dnhatn the tests are failing due to `EsqlActionTaskIT` - could you please take a look?

dnhatn commented:
The test failed because the query in the Lucene operator changed. It now executes:
```
{
  "exists" : {
    "field" : "pause_me",
    "boost" : 1.0
  }
}
```

Previously it was a match_all query. I think this can be a performance issue with the runtime fields.

dnhatn commented:
You can change `scriptPermits.release(PAGE_SIZE)`  to `scriptPermits.release(NUM_DOCS + PAGE_SIZE)` to make the test passed. However, that means there is a potential performance issue with runtime fields with this PR.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizer.java:
```diff
@@ -132,6 +166,77 @@ protected boolean skipExpression(Expression e) {
+    /**
+     * The vast majority of aggs ignore null entries - this rule adds a pushable filter for local aggs
+     * to filter this entries out to begin with.
+     * STATS x = min(a), y = sum(b)
```
I think this needs a grouping.

costin replied:
It doesn't. The filter only gets added if no grouping is present since it can filter them out if all the group values are null.

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -522,12 +553,25 @@ public void testQueryWithAggregation() {
+     *         \_FieldExtractExec[emp_no{f}#8]
+     *           \_EsQueryExec[test], query[{"exists":{"field":"emp_no","boost":1.0}}][_doc{f}#34], limit[], sort[] estimatedRowSize[8]
+     */
     public void testQueryWithAggAfterEval() {
```
If you could add one more test with two aggregations that show how the `bool` query looks like, that would be great.
For example `from test | stats agg_emp = sum(emp_no), min_s = min(salary)` where the push down should look something like `{"bool":{"should":[{"exists":{"field":"emp_no"}},{"exists":{"field":"salary"}}]}}`

astefan reviewed (approved):
LGTM
Left some minor testing and javadoc related comments.

costin committed:
Merge branch 'main' into fix/102787

costin committed:
Address feedback

costin commented:
Thanks folks - I've addressed the feedback and will break this PR in two: 1. to push the localaggregatefolding so it can be backported to 8.12 and the not-null inference to 8.13.

costin commented:
Closed in favor of #103673 (for 8.13) and #103706 (local aggregation for 8.12)
